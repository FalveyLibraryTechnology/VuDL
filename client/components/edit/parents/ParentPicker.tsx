import React, { useState } from "react";
import ObjectLoader from "../ObjectLoader";
import PidPicker from "../PidPicker";
import { useEditorContext } from "../../../context/EditorContext";
import { useFetchContext } from "../../../context/FetchContext";
import { getObjectLastChildPositionUrl, getParentUrl } from "../../../util/routes";

interface ParentPickerProps {
    pid: string;
}

const ParentPicker = ({ pid }: ParentPickerProps): React.ReactElement => {
    const {
        state: { objectDetailsStorage },
        action: {
            clearPidFromChildListStorage,
            removeFromObjectDetailsStorage,
            removeFromParentDetailsStorage,
            setSnackbarState,
        },
    } = useEditorContext();
    const {
        action: { fetchText },
    } = useFetchContext();
    const [selectedParentPid, setSelectedParentPid] = useState<string>("");
    const [position, setPosition] = useState<string>("");
    const [statusMessage, setStatusMessage] = useState<string>("");

    const loaded = Object.prototype.hasOwnProperty.call(objectDetailsStorage, selectedParentPid);
    const details = loaded ? objectDetailsStorage[selectedParentPid] : null;

    const showSnackbarMessage = (message: string, severity: string) => {
        setSnackbarState({
            open: true,
            message,
            severity,
        });
    };

    const errorCallback = (pid: string) => {
        showSnackbarMessage(`Cannot load details for ${pid}. Are you sure this is a valid PID?`, "error");
        setSelectedParentPid("");
    };

    const addParent = async () => {
        setStatusMessage("Saving...");
        const target = getParentUrl(pid, selectedParentPid);
        let result: string;
        try {
            result = await fetchText(target, { method: "PUT", body: position });
        } catch (e) {
            result = (e as Error).message ?? "Unexpected error";
        }
        if (result === "ok") {
            // Clear and reload the cached object and its parents, since these have now changed!
            removeFromObjectDetailsStorage(pid);
            removeFromParentDetailsStorage(pid);
            // Clear any cached lists belonging to the parent PID, because the
            // order has potentially changed!
            clearPidFromChildListStorage(selectedParentPid);
            showSnackbarMessage(`Successfully added ${pid} to ${selectedParentPid}`, "info");
        } else {
            showSnackbarMessage(result, "error");
        }
        setStatusMessage("");
    };

    const setToLastPosition = async () => {
        const target = getObjectLastChildPositionUrl(selectedParentPid);
        let result: string;
        try {
            result = await fetchText(target, { method: "GET" });
        } catch (e) {
            result = "0";
        }
        setPosition(parseInt(result) + 1);
    };

    const positionRequired = details && (details.sortOn ?? "") == "custom";
    const positionControl = positionRequired ? (
        <div>
            <label>
                Position: <input type="text" value={position} onChange={(e) => setPosition(e.target.value)} />
            </label>
            <button onClick={setToLastPosition}>Set to Last Position in Parent</button>
        </div>
    ) : null;

    let visibleMessage = "";
    if (positionRequired && position.length == 0) {
        visibleMessage = "Please enter a position.";
    } else if (!details) {
        visibleMessage = "Please select a valid PID.";
    } else {
        visibleMessage = statusMessage;
    }
    return (
        <>
            {selectedParentPid.length > 0 ? (
                <ObjectLoader pid={selectedParentPid} errorCallback={errorCallback} />
            ) : null}
            <PidPicker selected={selectedParentPid} setSelected={setSelectedParentPid} />
            <br />
            {positionControl}
            {visibleMessage.length == 0 ? <button onClick={addParent}>Add</button> : visibleMessage}
        </>
    );
};

export default ParentPicker;
