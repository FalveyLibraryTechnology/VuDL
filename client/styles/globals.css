:root,
[color-scheme="light"] {
    --body-bg: #fafbfc;
    --heading-color: #4d4d4d;
    --link-color: #008aed;

    --btn-bg: #485a6c;
    --btn-text: #ffffff;
    --btn-border: #275575;
    --btn-focus-bg: #3498db;

    --btn-primary-bg: #27ae60;
    --btn-primary-text: #ffffff;
    --btn-primary-border: #275575;
    --btn-primary-focus-bg: #155724;

    --btn-danger-bg: red;
    --btn-danger-text: #ffffff;
    --btn-danger-border: #a10705;
    --btn-danger-focus-bg: #a10705;

    --btn-logout-bg: #1abc9c;
    --btn-logout-text: #ffffff;
    --btn-logout-border: #16a085;
    --btn-logout-focus-bg: #1abc9c;

    --btn-toggle-bg: #7f8c8d;
    --btn-toggle-text: #ffffff;
    --btn-toggle-border: #34495e;
    --btn-toggle-focus-bg: #53626f;

    --group-border: #53626f;

    --input-border: #6a737d;
    --input-focus-border: #2188ff;

    --label-hover: #c8e1ff;

    --page-list-bg: #ffffff;
    --page-list-border: #c1c1c1;

    --thumbnail-bg: #e1e1e1;
    --thumbnail-focus-bg: #51aded;
    --thumbnail-number-bg: #ffffff;
    --thumbnail-number-text: #000000;
    --thumbnail-preview-border: #000000;

    --tree-item-focus-bg: #000000;
    --tree-item-focus-text: #ffffff;

    --shadow-gray: #c8e1ff;
}

[color-scheme="dark"] {
}
