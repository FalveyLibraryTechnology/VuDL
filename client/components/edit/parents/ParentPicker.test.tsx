import React from "react";
import { describe, beforeEach, expect, it, jest } from "@jest/globals";
import { mount, shallow } from "enzyme";
import { act } from "react-dom/test-utils";
import toJson from "enzyme-to-json";
import ParentPicker from "./ParentPicker";
import { waitFor } from "@testing-library/dom";

const mockUseEditorContext = jest.fn();
jest.mock("../../../context/EditorContext", () => ({
    useEditorContext: () => {
        return mockUseEditorContext();
    },
}));
const mockUseFetchContext = jest.fn();
jest.mock("../../../context/FetchContext", () => ({
    useFetchContext: () => {
        return mockUseFetchContext();
    },
}));
jest.mock("../ObjectLoader", () => () => "ObjectLoader");
let setSelected = (pid: string) => {
    // This placeholder function should never get called,
    // but we'll implement it as a safety fallback.
    throw new Error("Unexpected call: " + pid);
};
jest.mock("../PidPicker", () => (args) => {
    setSelected = args.setSelected;
    return "PidPicker";
});

describe("ParentPicker", () => {
    let editorValues;
    let fetchValues;
    const pid = "foo:123";
    const parentPid = "foo:122";
    beforeEach(() => {
        editorValues = {
            state: {
                objectDetailsStorage: {},
            },
            action: {
                clearPidFromChildListStorage: jest.fn(),
                removeFromObjectDetailsStorage: jest.fn(),
                removeFromParentDetailsStorage: jest.fn(),
                setSnackbarState: jest.fn(),
            },
        };
        fetchValues = {
            action: {
                fetchText: jest.fn(),
            },
        };
        mockUseEditorContext.mockReturnValue(editorValues);
        mockUseFetchContext.mockReturnValue(fetchValues);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it("renders correctly with no data loaded", () => {
        const wrapper = shallow(<ParentPicker pid={pid} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders correctly with a selected but unloaded parent", async () => {
        const wrapper = mount(<ParentPicker pid={pid} />);
        act(() => setSelected(parentPid));
        wrapper.update();
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders correctly with a selected, loaded, title-sorted parent", async () => {
        editorValues.state.objectDetailsStorage[parentPid] = {
            sortOn: "title",
        };
        const wrapper = mount(<ParentPicker pid={pid} />);
        act(() => setSelected(parentPid));
        wrapper.update();
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("adds a title-sorted parent", async () => {
        fetchValues.action.fetchText.mockResolvedValue("ok");
        editorValues.state.objectDetailsStorage[parentPid] = {
            sortOn: "title",
        };
        const wrapper = mount(<ParentPicker pid={pid} />);
        act(() => setSelected(parentPid));
        wrapper.update();
        await act(async () => {
            wrapper.find("button").simulate("click");
            await waitFor(() => expect(editorValues.action.setSnackbarState).toHaveBeenCalled());
        });
        expect(fetchValues.action.fetchText).toHaveBeenCalledWith(
            "http://localhost:9000/api/edit/object/foo%3A123/parent/foo%3A122",
            { body: "", method: "PUT" }
        );
        expect(editorValues.action.removeFromObjectDetailsStorage).toHaveBeenCalledWith(pid);
        expect(editorValues.action.removeFromParentDetailsStorage).toHaveBeenCalledWith(pid);
        expect(editorValues.action.clearPidFromChildListStorage).toHaveBeenCalledWith(parentPid);
        expect(editorValues.action.setSnackbarState).toHaveBeenCalledWith({
            message: "Successfully added foo:123 to foo:122",
            open: true,
            severity: "info",
        });
    });

    it("renders correctly with a selected, loaded, custom-sorted parent", async () => {
        editorValues.state.objectDetailsStorage[parentPid] = {
            sortOn: "custom",
        };
        const wrapper = mount(<ParentPicker pid={pid} />);
        act(() => setSelected(parentPid));
        wrapper.update();
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
