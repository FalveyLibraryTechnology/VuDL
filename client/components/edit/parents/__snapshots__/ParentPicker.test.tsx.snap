// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParentPicker handles object loading errors gracefully 1`] = `
[
  "PidPicker: {"selected":""}",
  <br />,
  "Please select a valid PID.",
]
`;

exports[`ParentPicker renders correctly with a selected but unloaded parent 1`] = `
[
  "ObjectLoader: {"pid":"foo:122"}",
  "PidPicker: {"selected":"foo:122"}",
  <br />,
  "Please select a valid PID.",
]
`;

exports[`ParentPicker renders correctly with a selected, loaded, custom-sorted parent 1`] = `
[
  "ObjectLoader: {"pid":"foo:122"}",
  "PidPicker: {"selected":"foo:122"}",
  <br />,
  <div>
    <label>
      Position: 
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <button
      onClick={[Function]}
    >
      Set to Last Position in Parent
    </button>
  </div>,
  "Please enter a position.",
]
`;

exports[`ParentPicker renders correctly with a selected, loaded, title-sorted parent 1`] = `
[
  "ObjectLoader: {"pid":"foo:122"}",
  "PidPicker: {"selected":"foo:122"}",
  <br />,
  <button
    onClick={[Function]}
  >
    Add
  </button>,
]
`;

exports[`ParentPicker renders correctly with no data loaded 1`] = `
[
  "PidPicker: {"selected":""}",
  <br />,
  "Please select a valid PID.",
]
`;
